Fix bug in Integration algorithm when using IncludePartialBins

DannyHindson opened the pull request:
**Description of work.**

When Integration is run with IncludePartialBins=True and integration limits that are either equal or close together the algorithm was returning the wrong answer. The partial bins logic wasn't working when the RangeLower and RangeUpper values were in between consecutive values in the x data

**To test:**

Run the script attached to the linked issue and observe that the algorithm now returns 0 in the calls with IncludePartialBins where the integration limits are the same

Fixes #32591. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Fix bug in Integration algorithm when using UsePartialBins

When Integration is run with UsePartialBins=True and integration
limits that are either equal or close together the algorithm was
returning the wrong answer. The partial bins logic wasn't working
when the RangeLower and RangeUpper values were in between consecutive
values in the x data

RichardWaiteSTFC commented on Framework/Algorithms/test/IntegrationTest.h:
```diff
@@ -146,15 +146,31 @@ class IntegrationTest : public CxxTest::TestSuite {
+    assertRangeWithPartialBins(input, 0.1, 0.5, {4., 6., 8.}, {1.265, 1.549, 1.788});
+  }
+
+  void testRangeWithPartialBinsAndEqualLimits() {
```
New tests look good, but is it also worthwhile having a test that asserts integration on a points and histogram data set that are equivalent give the same answer? It doesn't look like we have a test that does this but I may have missed it.

RichardWaiteSTFC reviewed (changes requested):
Works well - have tested on edge and points data and the bug is fixed.
Is it possible to update the docs for this algorithm - it says there 
`if the ranges given do not coincide with a bin boundary then the first bin boundary within the range is used.`
However, this only applies to `IncludePartialBins=False` is that correct?

DannyHindson commented:
> Works well - have tested on edge and points data and the bug is fixed. Is it possible to update the docs for this algorithm - it says there `if the ranges given do not coincide with a bin boundary then the first bin boundary within the range is used.` However, this only applies to `IncludePartialBins=False` is that correct?

Yes you're correct. I'll update. I think the comment was intended to emphasise the point that the algorithm doesn't do any rebinning but it's not accurate. If IncludePartialBins=True and the limits sit inside a bin then the algorithm calculates an extra fractional contribution.

DannyHindson committed:
Clarify misleading comment about limits

Correct statement about behaviour when limits not equal to bin boundaries

RichardWaiteSTFC reviewed (dismissed):
Thanks for fixing this and for clarifying the documentation, I agree changes to `setDistribution(true)` are beyond the scope of this PR.

DavidFair commented on Framework/Algorithms/src/Integration.cpp:
```diff
@@ -278,19 +279,35 @@ void Integration::exec() {
-        if (!is_distrib) {
-          fraction /= (next_bin - upper_bin);
+      } else {
+        if (distmin > 0) {
+          const double lower_bin = *lowit;
+          const double prev_bin = *(lowit - 1);
+          double fraction = (lower_bin - lowerLimit);
+          if (!is_distrib) {
+            fraction /= (lower_bin - prev_bin);
+          }
+          const MantidVec::size_type val_index = distmin - 1;
+          sumY += Y[val_index] * fraction;
+          const double eval = E[val_index];
+          sumE += eval * eval * fraction * fraction;
+          if (rebinned_input) {
+            sumF += Fmin * fraction;
+          }
```
Can we chunk this method and the other one into private methods. The `exec` of this file is starting to weigh in at 200+ LOC with the various nested if-else blocks

DannyHindson replied:
Just pulling out this ~10 lines into a private method will involve passing these 7 variables as parameters: lowit, lowerLimit, is_distrib, distmin, Y[], E[], Fmin and also returning 3 variables sumY, sumE and sumF. This feels like a slightly ugly bit to pull out.
Having had a look at the exec, perhaps the content of the parallel for loop could be pulled as some kind of IntegrateSpectrum method?

DavidFair replied:
Yeah that would help significantly too
This isn't a PR blocker, so if it's way too much work I can be swayed against

DannyHindson replied:
I've just pushed a commit with an IntegrateSpectrum method. See what you think. I can always revert to the previous commit if this is getting too big a change but seems OK?

DavidFair commented on Framework/Algorithms/src/Integration.cpp:
```diff
@@ -278,19 +279,35 @@ void Integration::exec() {
-        sumE += eval * eval * fraction * fraction;
-        if (rebinned_input) {
-          sumF += Fmax * fraction;
+        if (highit < X.end() - 1) {
+          const double upper_bin = *highit;
+          const double next_bin = *(highit + 1);
+          double fraction = (upperLimit - upper_bin);
+          if (!is_distrib) {
+            fraction /= (next_bin - upper_bin);
+          }
+          sumY += Y[distmax] * fraction;
+          const double eval = E[distmax];
+          sumE += eval * eval * fraction * fraction;
+          if (rebinned_input) {
+            sumF += Fmax * fraction;
+          }
```
Same with this too

DavidFair reviewed (changes requested):
Some minor readability nits, otherwise I'm happy to merge this

DannyHindson committed:
Pull out per spectrum integration into separate method

Exec method was getting too large so pull out the per spectrum
logic into its own private method

DavidFair commented on Framework/Algorithms/inc/MantidAlgorithms/Integration.h:
```diff
@@ -69,6 +69,9 @@ class MANTID_ALGORITHMS_DLL Integration : public API::Algorithm {
   /// Create the outputworkspace
   API::MatrixWorkspace_sptr getOutputWorkspace(API::MatrixWorkspace_sptr inWS, const int minSpec, const int maxSpec);
   std::map<std::string, std::string> validateInputs() override;
+  void integrateSpectrum(const API::ISpectrum &inSpec, API::ISpectrum &outSpec, const std::vector<double> *Fin,
```
LGTM, my only nits with this are:
- Can allocate a vector on the stack and pass in by ref and just check for .empty as the sentinel
- Clang-tidy will complain about `const` scalar types which are passed by copy, can we drop const from double and bool?

DannyHindson replied:
No problem about removing the `const`s. Re your first point, I could do this for Fin but I think I have to leave Fout as a pointer because I need to update the data inside rebinned_output->dataF(). So seemed more symmetrical to leave Fin and Fout as pointers? Knew you'd like a bit of raw pointer action...

DavidFair replied:
That sounds good to me, we'll stick with the raw pointer squad so we can get this out of the way for the weekend :laughing:

DavidFair reviewed (dismissed):
LGTM I'll :shipit: once the tests pass

DavidFair commented:
It looks like you're missing the header for `API::ISpectrum` which is why it's falling over outside of MSVC

RichardWaiteSTFC reviewed (approved):
All builds passing now!

DavidFair reviewed (approved):
Thanks all :+1:
